src/coreclr/vm/amd64/AsmHelpers.asm

```diff
@@ -509,6 +509,56 @@ LEAF_ENTRY ThisPtrRetBufPrecodeWorker, _TEXT
     jmp METHODDESC_REGISTER
 LEAF_END ThisPtrRetBufPrecodeWorker, _TEXT
 
+;;
+;; Prologue of all funclet calling helpers (CallXXXXFunclet)
+;;
+FUNCLET_CALL_PROLOGUE macro localsCount, alignStack
+        PUSH_CALLEE_SAVED_REGISTERS
+
+        arguments_scratch_area_size = 20h
+        xmm_save_area_size = 10 * 10h ;; xmm6..xmm15 save area
+        stack_alloc_size = arguments_scratch_area_size + localsCount * 8 + alignStack * 8 + xmm_save_area_size
+        rsp_offsetof_arguments = stack_alloc_size + 8*8h + 8h
+        rsp_offsetof_locals = arguments_scratch_area_size + xmm_save_area_size
+
+        alloc_stack     stack_alloc_size
+
+        vzeroupper
```

Comment:
Why do we need the vzeroupper?

Reply:
Right, it is clearing the AVX state. The JIT does it as part of pushing/polling callee saved registers. We are pushing/polling callee saved registers unconditionally in the asm helper, so it makes sense to do `vzeroupper` there as well.

The JIT uses `vzeroupper` liberally. It is very cheap instruction (it just sets a flag), but the perf consequence of missing it are very severe (5x slowdown).

I have added a comment and also cleaned up where `vzeroupper` is added in the asmhelper. It is always done as part of the PROLOG/EPILOG macros now.